File: android/src/main/kotlin/dev/brewkits/native_workmanager/workers/DartCallbackWorker.kt
package dev.brewkits.native_workmanager.workers

import android.content.Context
import android.util.Log
import dev.brewkits.kmpworkmanager.background.domain.AndroidWorker
import dev.brewkits.kmpworkmanager.background.domain.WorkerResult
import dev.brewkits.native_workmanager.engine.FlutterEngineManager
import org.json.JSONObject

/**
 * Android worker that executes Dart callbacks in background.
 *
 * This worker:
 * 1. Receives a callback handle (from PluginUtilities.getCallbackHandle)
 * 2. Starts/reuses Flutter Engine
 * 3. Invokes the Dart callback via MethodChannel
 * 4. Returns the callback result
 *
 * Performance characteristics:
 * - Cold start (first task): 500-1000ms
 * - Warm start (engine cached): 100-200ms
 * - RAM usage: ~30-50MB while engine is alive
 *
 * Input JSON format:
 * ```json
 * {
 *   "callbackId": "myCallback",        // For logging/debugging
 *   "callbackHandle": 12345678,        // ✅ Serializable handle (REQUIRED)
 *   "input": "{\"key\": \"value\"}",   // Optional JSON string input
 *   "autoDispose": true                // Optional: Kill engine immediately after completion (default: false)
 * }
 * ```
 *
 * @see FlutterEngineManager
 */
class DartCallbackWorkerWrapper(
    private val context: Context
) : AndroidWorker {

    companion object {
        private const val TAG = "DartCallbackWorker"
    }

    /**
     * Execute the Dart callback.
     *
     * This method is called by WorkManager in a background thread.
     * It's blocking but that's OK since WorkManager handles threading.
     *
     * @param input JSON string containing callback handle and optional input data
     * @return WorkerResult indicating success/failure (data from Dart callback)
     */
    override suspend fun doWork(input: String?): WorkerResult {
        return try {
            Log.d(TAG, "DartCallbackWorker started")

            if (input == null || input.isEmpty()) {
                Log.e(TAG, "Input is null or empty")
                return WorkerResult.Failure("Input is null or empty")
            }

            // Parse input JSON
            val json = try {
                JSONObject(input)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to parse input JSON: $input", e)
                return WorkerResult.Failure("Failed to parse input JSON: ${e.message}")
            }

            // Extract callback handle (REQUIRED)
            val callbackHandle = try {
                json.getLong("callbackHandle")
            } catch (e: Exception) {
                Log.e(TAG, "Missing or invalid callbackHandle in input: $input", e)
                Log.e(TAG, "Input JSON keys: ${json.keys().asSequence().toList()}")
                return WorkerResult.Failure("Missing or invalid callbackHandle")
            }

            // Extract callbackId (for logging only)
            val callbackId = json.optString("callbackId", "unknown")

            // Extract optional input data
            val callbackInput = json.optString("input", null)

            // ✅ NEW: Extract autoDispose flag (default: false)
            val autoDispose = json.optBoolean("autoDispose", false)

            Log.d(TAG, "Executing callback: $callbackId (handle: $callbackHandle, autoDispose: $autoDispose)")

            // Execute Dart callback via FlutterEngineManager
            // ✅ Pass callbackHandle (not callbackId) to enable cross-isolate execution
            val result = FlutterEngineManager.executeDartCallback(
                context = context,
                callbackHandle = callbackHandle,  // ✅ Serializable handle
                input = callbackInput,
                timeoutMs = 5 * 60 * 1000L, // 5 minutes timeout
                disposeImmediately = autoDispose // ✅ NEW: Aggressive disposal flag
            )

            Log.d(TAG, "Dart callback completed: $callbackId, result: $result")

            if (result) {
                WorkerResult.Success(message = "Dart callback executed: $callbackId")
            } else {
                WorkerResult.Failure("Dart callback returned false: $callbackId")
            }

        } catch (e: Exception) {
            Log.e(TAG, "Error in DartCallbackWorker", e)
            WorkerResult.Failure(e.message ?: "Unknown error", shouldRetry = true)
        }
    }
}

File: ios/Classes/workers/DartCallbackWorker.swift
import Foundation

/// Dart callback worker for iOS.
///
/// Executes Dart code in a background isolate using FlutterEngineManager.
/// Provides Mode 2 execution (Dart workers) for complex business logic
/// that requires Flutter plugins or Dart-specific functionality.
///
/// **Configuration JSON:**
/// ```json
/// {
///   "callbackId": "myCallback",        // For logging/debugging
///   "callbackHandle": 12345678,        // ✅ Serializable handle (REQUIRED)
///   "input": "{\"key\":\"value\"}",    // Optional: JSON input string
///   "timeoutMs": 300000,               // Optional: Timeout (default: 5 minutes)
///   "autoDispose": true                // Optional: Kill engine immediately after completion (default: false)
/// }
/// ```
///
/// **Performance:**
/// - Cold start (first execution): 500-1000ms + callback time
/// - Warm start (cached engine): 100-200ms + callback time
/// - Memory: 30-50MB when engine active
///
/// **Use Cases:**
/// - Complex Dart business logic
/// - Access to Flutter plugins (database, etc.)
/// - State management operations
/// - Custom data processing
class DartCallbackWorker: IosWorker {

    // MARK: - Singleton

    static let shared = DartCallbackWorker()

    private init() {}

    // MARK: - Configuration

    private static let defaultTimeoutMs: Int64 = 300_000 // 5 minutes

    struct Config: Codable {
        let callbackId: String       // For logging/debugging
        let callbackHandle: Int64    // ✅ Serializable handle (REQUIRED)
        let input: String?
        let timeoutMs: Int64?
        let autoDispose: Bool?       // ✅ NEW: Aggressive disposal flag

        var timeoutSeconds: TimeInterval {
            TimeInterval((timeoutMs ?? DartCallbackWorker.defaultTimeoutMs) / 1000)
        }

        var shouldAutoDispose: Bool {
            autoDispose ?? false
        }
    }

    // MARK: - Worker Implementation

    func doWork(input: String?) async throws -> WorkerResult {
        guard let input = input, !input.isEmpty else {
            print("DartCallbackWorker: Error - Empty or null input")
            return WorkerResult.failure(message: "Empty or null input")
        }

        // Parse configuration
        guard let data = input.data(using: .utf8) else {
            print("DartCallbackWorker: Error - Invalid UTF-8 encoding")
            return WorkerResult.failure(message: "Invalid input encoding")
        }

        let config: Config
        do {
            config = try JSONDecoder().decode(Config.self, from: data)
        } catch {
            print("DartCallbackWorker: Error parsing JSON config: \(error)")
            return WorkerResult.failure(message: "Error parsing JSON config: \(error.localizedDescription)")
        }

        print("DartCallbackWorker: Executing callback '\(config.callbackId)' (handle: \(config.callbackHandle), autoDispose: \(config.shouldAutoDispose))")

        let startTime = Date()
        let engineWasAlive = FlutterEngineManager.shared.isEngineAlive

        // ✅ FIXED: Execute callback via FlutterEngineManager with callbackHandle
        // The callbackHandle enables cross-isolate execution
        do {
            let result = try await FlutterEngineManager.shared.executeDartCallback(
                callbackHandle: config.callbackHandle,  // ✅ Pass serializable handle
                input: config.input,
                timeoutSeconds: config.timeoutSeconds,
                disposeImmediately: config.shouldAutoDispose // ✅ NEW: Aggressive disposal flag
            )

            let executionTime = Date().timeIntervalSince(startTime)

            if result {
                if engineWasAlive {
                    print("DartCallbackWorker: Success (warm start) - \(Int(executionTime * 1000))ms")
                } else {
                    print("DartCallbackWorker: Success (cold start) - \(Int(executionTime * 1000))ms")
                }
                return WorkerResult.success(message: "Callback returned true")
            } else {
                print("DartCallbackWorker: Failed - Callback returned false")
                return WorkerResult.failure(message: "Callback returned false")
            }
        } catch {
            let executionTime = Date().timeIntervalSince(startTime)
            print("DartCallbackWorker: Error - \(error.localizedDescription) (\(Int(executionTime * 1000))ms)")
            return WorkerResult.failure(message: "Execution error: \(error.localizedDescription)")
        }
    }
}
